### PR TITLE
Clarify recommended network setup

### LINKED ISSUE #2438: Clarify recommended setup
_EDIT as to provide guidance to new users - Matrix devs official answer [#1](https://github.com/matrix-org/synapse/issues/2438#issuecomment-327424711) and [#2](https://github.com/matrix-org/synapse/issues/2438#issuecomment-327425464)_

The self-signed certificate of synapse should NOT be replaced and port 8448 should only be used for federation (server traffic) and directly exposed publicly.
For clients connections, a reverse proxy should be reachable publicly with a regular certificate (e.g. Let's Encrypt) on port 443 that goes to the port 8008 of synapse.

**Public TCP 8448**: Servers, original generated self-signed certificate, directly exposing port TCP 8448 of synapse
**Public TCP 443**: Clients, regular certificate (e.g. Let's Encrypt), reversed proxy to port TCP 8008 of synapse

---

The community is currently discussing what is the official recommended setup for synapse in terms of ports and reverse proxying. Here are the three visions:
1. Use a regular (e.g. LTE) certificate on port 8448 and expose that for clients and servers, replacing the self-signed generated one.
2. Use a regular certificate on 443 reversed proxy to 8008 and same cert on 8448, replacing the self-signed generated one.
3. Use a regular certificate on 443 reversed proxy to 8008 and keep the self-signed generated one on 8448

My personal believe is to do go for 3) but it seems many users in the community for for 1) or 2), which I believe is due to misleading documentation for synapse.

Why I think 1) and 2) occurs due to misleading doc:
First mention of 8008 and 8448 is at [Configuring synapse](https://github.com/matrix-org/synapse/blob/c0c31656ff379c0a2ec54e6af450f12bb2f6c2fb/README.rst#configuring-synapse), last paragraph which gives the first misleading advise.
It says 8448 has a self-signed certificate which is fine to test with clients but should be replaced in production but also state that it's fine for Federation. So if you want to connect clients to 8448, you should put a proper certificate - yet doesn't instruct to connect clients to it outside of a testing scope.

The next bit comes with [Connecting to Synapse from a client](https://github.com/matrix-org/synapse/blob/c0c31656ff379c0a2ec54e6af450f12bb2f6c2fb/README.rst#connecting-to-synapse-from-a-client) which says again that the easiest way to try a synapse install is to connect to 8448 but totally ignores the fact it's a self-signed and will be a problem. So you're made to believe you should have changed the certificate.

Then we come to [Using a reverse proxy with Synapse](https://github.com/matrix-org/synapse/blob/c0c31656ff379c0a2ec54e6af450f12bb2f6c2fb/README.rst#using-a-reverse-proxy-with-synapse) which says it will work nicely with clients as you expose 443 and continues by saying clients and servers don't necessarily need to connect to the same port, says and shows a nginx config of 443->8008 for clients.

Finally, several generic points:
- Changing the certificate only matters for clients. Servers don't require anything specific.
- 8448 doesn't work out of the box for clients (due to self-signed) but does for federation
- 443 to a unencrypted 8xxx port is the typical setup with TLS offloading
- 8008 is enabled by default in the config. If 1) was intended, why not just keep 8448?
- if you use 443 for clients, you don't need to give the port when configuring, just put https://example.org - easy for users, easy for admins
- Reverse proxy 443 to 8448 with self-signed certificate might not work out of the box, so you would go for 8008 (and double TLS/SSL would not make much sense)
- synapse README never instructs or even recommends to use 8448 for clients in production, only for testing
- going for 2) means managing two configuration files and two copies of the same certs after each renewal
- The [Server-Server Spec](https://matrix.org/docs/spec/server_server/unstable.html#retrieving-server-keys) talks about caching results of looking up certificates and signing keys, which could cause downtime depending on homeservers implementation/configuiration?
---
**Could the Matrix devs please clearly state what is the recommended setup?**

### PR DESCRIPTION
**Pull request by maxidor**
Because of the constant flow of people screwing up with the network setup and TLS certificates, here is an improved README to clarify various points.
Should fix #2438

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -200,19 +200,21 @@ different. See `the spec`__ for more information on key management.)
...
 certificate is fine for `Federation`_). You can do so by changing
 ``tls_certificate_path``, ``tls_private_key_path`` and ``tls_dh_params_path``
-in ``homeserver.yaml``; alternatively, you can use a reverse-proxy, but be sure
-to read `Using a reverse proxy with Synapse`_ when doing so.
+in ``homeserver.yaml``;
 
 Apart from port 8448 using TLS, both ports are the same in the default
 configuration.
 
+See https://github.com/matrix-org/synapse/issues/2438 for the recommended
```
Comment (richvdh) [inline README.rst]: Do we need this? I don't think it adds much other than confusion here.

File: README.rst [hunk 2]
```diff
@@ -200,19 +200,21 @@ different. See `the spec`__ for more information on key management.)
...
-another certificate for production purposes. (Note that a self-signed
+configured without TLS; it should be behind a reverse proxy for TLS/SSL
+termination on port 443 which in turn should be used for clients. Port 8448
+is configured to use TLS with a self-signed certificate. If you would like
+to do initial test with a client without having to setup a reverse proxy,
+you can temporarly use another certificate. (Note that a self-signed
 certificate is fine for `Federation`_). You can do so by changing
 ``tls_certificate_path``, ``tls_private_key_path`` and ``tls_dh_params_path``
-in ``homeserver.yaml``; alternatively, you can use a reverse-proxy, but be sure
-to read `Using a reverse proxy with Synapse`_ when doing so.
```
Comment (richvdh) [inline README.rst]: Can we keep the link to the 'reverse-proxy' section?

File: README.rst [hunk 3]
```diff
@@ -593,8 +601,9 @@ you to run your server on a machine that might not have the same name as your
 domain name. For example, you might want to run your server at
 ``synapse.example.com``, but have your Matrix user-ids look like
 ``@user:example.com``. (A SRV record also allows you to change the port from
-the default 8448. However, if you are thinking of using a reverse-proxy, be
-sure to read `Reverse-proxying the federation port`_ first.)
+the default 8448. However, if you are thinking of using a reverse-proxy on the
+federation port, which is highly not recommended, be sure to read
```
Comment (richvdh) [inline README.rst]: s/highly//. If nothing else, you'll need a reverse-proxy if you have a synapse with split workers, so this is a bit strong.

**Comment by maxidor** [comment]
@richvdh Improved following your feedback

**Pull request merged**